compiler/modules/types/core.bal

```diff
@@ -960,23 +960,31 @@ function bddMappingAtomicType(Env env, Bdd bdd, MappingAtomicType top) returns M
 // for when T is a subtype of mapping, and K is either `string` or a singleton string.
 // This is what Castagna calls projection.
 // We will extend this to allow `key` to be a SemType, which will turn into a StringSubtype.
-public function mappingMemberType(Context cx, SemType t, string? k = ()) returns SemType {
+public function mappingMemberType(Context cx, SemType t, SemType k = STRING) returns SemType {
     if t is UniformTypeBitSet {
         return (t & MAPPING) != 0 ? TOP : NEVER;
     }
     else {
-        return union(bddMappingMemberType(cx, <Bdd>getComplexSubtypeData(t, UT_MAPPING_RO), k, TOP),
-                     bddMappingMemberType(cx, <Bdd>getComplexSubtypeData(t, UT_MAPPING_RW), k, TOP));
+        StringSubtype? key = ();
+        if k is ComplexSemType {
+            key = <StringSubtype>getComplexSubtypeData(k, UT_STRING);
+        }
+        return union(bddMappingMemberType(cx, <Bdd>getComplexSubtypeData(t, UT_MAPPING_RO), key, TOP),
+                     bddMappingMemberType(cx, <Bdd>getComplexSubtypeData(t, UT_MAPPING_RW), key, TOP));
     }
 }
 
-public function mappingMemberRequired(Context cx, SemType t, string k) returns boolean {
+public function mappingMemberRequired(Context cx, SemType t, SemType k) returns boolean {
+    if k !is ComplexSemType {
```

Comment:
Can we combine this with the next test?

Reply:
I assumed that we might not get the narrowing, I was wrong.